tests/integration/deprecated_syscalls_tests.rs

```diff
@@ -491,3 +492,103 @@ async fn test_syscall_get_contract_address_cairo0(
+
+    let tx_version = TransactionVersion::ZERO;
+    let keys = vec![stark_felt!(2019_u16), stark_felt!(2020_u16)];
+    let data = vec![stark_felt!(2021_u16), stark_felt!(2022_u16), stark_felt!(2023_u16)];
+    let entrypoint_args = &[
+        vec![stark_felt!(u128::try_from(keys.len()).unwrap())],
+        keys,
+        vec![stark_felt!(u128::try_from(data.len()).unwrap())],
+        data,
+    ]
+    .concat();
+
+    let mut nonce_manager = NonceManager::default();
+    let tx = test_utils::account_invoke_tx(invoke_tx_args! {
+        max_fee,
+        sender_address: sender_address,
+        calldata: create_calldata(contract_address, "test_emit_event", entrypoint_args),
+        version: tx_version,
+        nonce: nonce_manager.next(sender_address),
+    });
+
+    let txs = vec![tx];
+
+    let (_pie, _os_output) = execute_txs_and_run_os(
+        initial_state.cached_state,
+        block_context.clone(),
+        txs,
+        initial_state.cairo0_compiled_classes,
+        initial_state.cairo1_compiled_classes,
+    )
+    .await
+    .expect("OS run failed");
+}
+
+#[rstest]
+// We need to use the multi_thread runtime to use task::block_in_place for sync -> async calls.
+#[tokio::test(flavor = "multi_thread", worker_threads = 1)]
+async fn test_syscall_send_message_to_l1_cairo0(
+    #[future] initial_state_cairo0: StarknetTestState,
+    block_context: BlockContext,
+    max_fee: Fee,
+) {
+    let initial_state = initial_state_cairo0.await;
+
+    let sender_address = initial_state.cairo0_contracts.get("account_with_dummy_validate").unwrap().address;
+    let test_contract = initial_state.cairo0_contracts.get("test_contract").unwrap();
+    let contract_address = test_contract.address;
+
+    let tx_version = TransactionVersion::ZERO;
+
+    let address = 1234_u16;
+    let to_address = stark_felt!(address);
+
+    let entrypoint_args = &[vec![to_address]].concat();
+
+    let mut nonce_manager = NonceManager::default();
+    let tx = test_utils::account_invoke_tx(invoke_tx_args! {
+        max_fee,
+        sender_address: sender_address,
+        calldata: create_calldata(contract_address, "send_message", entrypoint_args),
+        version: tx_version,
+        nonce: nonce_manager.next(sender_address),
+    });
+
+    let txs = vec![tx];
+
+    let (_pie, os_output) = execute_txs_and_run_os(
+        initial_state.cached_state,
+        block_context.clone(),
+        txs,
+        initial_state.cairo0_compiled_classes,
+        initial_state.cairo1_compiled_classes,
+    )
+    .await
+    .expect("OS run failed");
+    let output: Vec<BigUint> =
+        os_output.messages_to_l1.iter().map(|msg| BigUint::from_bytes_le(&msg.to_bytes_le())).collect();
+
+    // assert the address is returned
+    assert!(&output.contains(&address.try_into().unwrap()));
+
+    // assert the payload
+    for expected_payload in [12u16, 34] {
+        assert!(&output.contains(&expected_payload.try_into().unwrap()));
+    }
```

Comment:
As discussed, this is too simplistic and could lead to error. Now that we know that the message looks like:
1. Sender address
2. `to_address`
3. payload length
4+. payload

I want you to test that you get exactly the expected message.